Dashboard docs refactor (#674)
dotnet/docs-aspire: docs/fundamentals/dashboard/security-considerations.md

---
title: .NET Aspire dashboard security considerations
description: Security considerations for running the .NET Aspire dashboard
ms.date: 03/13/2024
ms.topic: reference
---

# Security considerations for running the .NET Aspire dashboard

The [.NET Aspire dashboard](overview.md) offers powerful insights to your apps. The dashboard displays information about resources, including their configuration, console logs and in-depth telemtry.

Data displayed in the dashboard can be sensitive. For example, configuration can include secrets in environment variables, and telemetry can include sensitive runtime data. Care should be taken to secure access to the dashboard.

## Scnearios for running the dashboard

> The dashboard can be run in different scenarios, such as being automatically starting by Aspire tooling, or as a standalone application that is separate from other Aspire components. Steps to secure the dashboard depend on how it is being run.

Reviewer: ```suggestion
The dashboard can be run in different scenarios, such as being automatically starting by .NET Aspire tooling, or as a standalone application that is separate from other .NET Aspire components. Steps to secure the dashboard depend on how it's being run.
```

> ### Aspire tooling

Reviewer: ```suggestion
### .NET Aspire tooling
```

> The dashboard is automatically started when an Aspire app host is run. The dashboard is secure by default when run from Aspire tooling:

Reviewer: ```suggestion
The dashboard is automatically started when an .NET Aspire app host is run. The dashboard is secure by default when run from .NET Aspire tooling:
```

- Transport is secured with HTTPS. Using HTTPS is configured by default in `launchSettings.json`. The launch profile includes `https` addresses in `applicationUrl` and `DOTNET_DASHBOARD_OTLP_ENDPOINT_URL` values.
- Browser frontend authenticated with a browser token.
- Incoming telemetry authenticated with an API key.

HTTPS in the dashboard uses the ASP.NET Core development certificate. The certificate must be trusted for the dashboard to work correctly. The steps required to trust the development cert are different depending on the machine's operating system:

> - [Trust the ASP.NET Core HTTPS development certificate on Windows and macOS](https://learn.microsoft.com/aspnet/core/security/enforcing-ssl#trust-the-aspnet-core-https-development-certificate-on-windows-and-macos)

Reviewer: ```suggestion
- [Trust the ASP.NET Core HTTPS development certificate on Windows and macOS](/aspnet/core/security/enforcing-ssl#trust-the-aspnet-core-https-development-certificate-on-windows-and-macos)
```

> - [Trust HTTPS certificate on Linux](https://learn.microsoft.com/aspnet/core/security/enforcing-ssl#trust-https-certificate-on-linux)

Reviewer: ```suggestion
- [Trust HTTPS certificate on Linux](/aspnet/core/security/enforcing-ssl#trust-https-certificate-on-linux)
```

> There are scenarios where you might want to allow an unsecured transport. The dashboard can run without HTTPS from the Aspire app host by configuring the `ASPIRE_ALLOW_UNSECURED_TRANSPORT` setting to `true`. For more information, see [Allow unsecured transport in .NET Aspire](../../troubleshooting/allow-unsecure-transport.md).

Reviewer: ```suggestion
There are scenarios where you might want to allow an unsecured transport. The dashboard can run without HTTPS from the .NET Aspire app host by configuring the `ASPIRE_ALLOW_UNSECURED_TRANSPORT` setting to `true`. For more information, see [Allow unsecured transport in .NET Aspire](../../troubleshooting/allow-unsecure-transport.md).
```

### Standalone mode

> The dashboard is shipped as a docker image and can be without the rest of .NET Aspire. When the dashboard is launched in standalone mode, it defaults to a mix of secure and unsecured settings.

Reviewer: ```suggestion
The dashboard is shipped as a Docker image and can be without the rest of .NET Aspire. When the dashboard is launched in standalone mode, it defaults to a mix of secure and unsecured settings.
```

API key authentication can be enabled on the telemetry endpoint with some additional configuration:

```bash
docker run --rm -it -p 18888:18888 -p 4317:18889 -d --name aspire-dashboard \
    -e DASHBOARD__OTLP__AUTHMODE='ApiKey' \
    -e DASHBOARD__OTLP__PRIMARYAPIKEY='{MY_APIKEY}' \
    mcr.microsoft.com/dotnet/nightly/aspire-dashboard:8.0.0-preview.6
```

The proceding Docker command:

- Starts the .NET Aspire dashboard image and exposes OTLP endpoint as port 4317
- Configures the OTLP endpoint to use `ApiKey` authentication. This requires that incoming telemetry has a valid `x-otlp-api-key` header value.
- Configures the expected API key. `{MY_APIKEY}` in the example value should be replaced with a real API key. The API key can be any text, but a value with at least 128 bits of entropy is recommended.

When API key authentication is configured, the dashboard validates incoming telemetry has a required API key. Apps that send the dashboard telemetry must be configured to send the API key. This can be configured in .NET with `OtlpExporterOptions.Headers`:

```csharp

> builder.Services.Configure<OtlpExporterOptions>(o => o.Headers = $"x-otlp-api-key={MY_APIKEY}");

Reviewer: Help prevent horizontal scrolling on narrow screens.

```suggestion
builder.Services.Configure<OtlpExporterOptions>(
    o => o.Headers = $"x-otlp-api-key={MY_APIKEY}");
```
